src/GovUk.Education.ExploreEducationStatistics.Common/Cache/BlobCacheAttribute.cs

```diff
@@ -0,0 +1,88 @@
+#nullable enable
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using GovUk.Education.ExploreEducationStatistics.Common.Cache.Interfaces;
+using GovUk.Education.ExploreEducationStatistics.Common.Extensions;
+using GovUk.Education.ExploreEducationStatistics.Common.Services.Interfaces;
+
+namespace GovUk.Education.ExploreEducationStatistics.Common.Cache
+{
+    public class BlobCacheAttribute : CacheAttribute
+    {
+        private static Dictionary<string, IBlobCacheService> Services { get; set; } = new();
+
+        protected override Type BaseKey => typeof(IBlobCacheKey);
+
+        /// <summary>
+        /// Specify a service to use <see cref="Services"/>.
+        /// Otherwise, we use the first registered service.
+        /// </summary>
+        public string? ServiceName { get; set; }
+
+        public BlobCacheAttribute(Type key) : base(key)
+        {
+        }
+
+        public static void AddService(string name, IBlobCacheService service)
+        {
+            Services[name] = service;
+        }
+        public static void RemoveService(string name)
+        {
+            Services.Remove(name);
+        }
+
+        public static void ClearServices()
+        {
+            Services.Clear();
+        }
+
+        public override async Task<object?> Get(ICacheKey cacheKey, Type returnType)
+        {
+            if (cacheKey is IBlobCacheKey key)
```

Comment:
I think this would be obvious if I switched to browsing with IDE on this branch, but why is this cacheKey a parameter and not the key that this is constructed with, which is already validated for being a type of IBlobCacheKey in the protected constructor of CacheAttribute?

Reply:
So the main thing is that we're inheriting from `CacheAttribute` that defines abstract `Get` and `Set` methods. As we want to override these, we can't then redefine them in `BlobCacheAttribute` because parameters can't be covariant without using generics.

You might be thinking this would simply work by adding a type parameter on `CacheAttribute`, but unfortunately nope! We can't do this because attributes can't have type parameters unfortunately :cry:

So the only option is to override the `CacheAttribute` methods and just do a type check once inside the method basically!

Reply:
Thanks I see that now, and yes that's a pain! 🙂